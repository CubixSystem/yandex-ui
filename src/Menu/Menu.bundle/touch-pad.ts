import { FC } from 'react';
import { compose, composeU } from '@bem-react/core';

// base
import { IMenuProps as IMenuTouchPadProps, Menu as MenuTouchPad } from '../Menu@touch-pad';
// _size
import { withSizeM } from '../_size/Menu_size_m';
import { withSizeS } from '../_size/Menu_size_s';
// _theme
import { withThemeNormal } from '../_theme/Menu_theme_normal';
// _view
import { withViewDefault } from '../_view/Menu_view_default';
// _width
import { withWidthAuto } from '../_width/Menu_width_auto';
import { withWidthMax } from '../_width/Menu_width_max';

export * from '../Menu@touch-pad';

export interface IMenuProps extends IMenuTouchPadProps {
    size?: 'm' | 's';
    view?: 'default';
    width?: 'auto' | 'max';
    theme?: 'normal';
}

export const Menu = compose(
    composeU(withSizeM, withSizeS),
    composeU(withWidthAuto, withWidthMax),
    withViewDefault,
    withThemeNormal,
)(MenuTouchPad) as FC<IMenuProps>;
