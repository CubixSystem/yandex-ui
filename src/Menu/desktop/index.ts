export * from '../Menu@desktop';
export * from '../_size/Menu_size_m';
export * from '../_size/Menu_size_s';
export * from '../_theme/Menu_theme_normal@desktop';
export * from '../_view/Menu_view_default@desktop';
export * from '../_width/Menu_width_auto';
export * from '../_width/Menu_width_max';
