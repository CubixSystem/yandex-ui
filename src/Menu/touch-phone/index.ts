export * from '../Menu@touch-phone';
export * from '../_jsxContent/Menu_jsxContent';
export * from '../_size/Menu_size_m';
export * from '../_size/Menu_size_s';
export * from '../_theme/Menu_theme_normal';
export * from '../_view/Menu_view_default';
export * from '../_width/Menu_width_auto';
export * from '../_width/Menu_width_max';
export * from '../Menu.hooks/useMenuItem';
