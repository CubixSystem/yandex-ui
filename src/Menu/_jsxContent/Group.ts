import { ReactNode } from 'react';

import { CollectionComponent } from '../../useCollection';
import { ItemGroup } from '../Menu';

export interface GroupProps {
    /**
     * Заголовок группы
     */
    label?: string;

    /**
     * Дополнительный класс
     */
    className?: string;

    /**
     * Набор элементов в группе
     */
    children?: ReactNode;
}

export const Group: CollectionComponent<GroupProps, ItemGroup> = () => {
    return null;
};

Group.getCollectionNode = function getComponetNode(props, context): ItemGroup {
    // eslint-disable-next-line camelcase
    const { children, label, ...unsafe_props } = props;

    return {
        title: label,
        items: context.createCollection(props.children),
        unsafe_props,
    };
};
