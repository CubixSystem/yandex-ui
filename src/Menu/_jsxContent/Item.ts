import { ReactNode } from 'react';

import { CollectionComponent } from '../../useCollection';
import { ItemSimple } from '../Menu';

export interface ItemProps {
    /**
     * Значение, которое соответствует пункту меню. Это значение передается в обработчик, который указан в `onChange`
     */
    value?: any;

    /**
     * Неактивное состояние элемента
     */
    disabled?: boolean;

    /**
     * Идентификатор компонента
     */
    id?: string;

    /**
     * Дополнительный класс
     */
    className?: string;

    /**
     * Содержимое элемента
     */
    children?: ReactNode;
}

export const Item: CollectionComponent<ItemProps, ItemSimple> = () => {
    return null;
};

Item.getCollectionNode = function getComponetNode(props): ItemSimple {
    // eslint-disable-next-line camelcase
    const { children, value, disabled, id, ...unsafe_props } = props;

    return {
        id,
        disabled,
        value,
        content: children,
        unsafe_props,
    };
};
