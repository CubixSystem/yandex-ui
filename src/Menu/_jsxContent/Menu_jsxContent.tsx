import React, { FC, ReactNode, ComponentType } from 'react';

import { getDisplayName } from '../../lib/getDisplayName';
import { useCollection } from '../../useCollection';
import { IMenuProps, MixedItem } from '../Menu';

type WithItemsProps = Pick<IMenuProps, 'items'>;

export interface MenuJsxContentProps {
    /**
     * Список пунктов меню в виде JSX
     */
    children?: ReactNode;

    /**
     * Список пунктов меню в виде массива
     */
    items?: MixedItem[];
}

/**
 * Модификатор, отвечающий за содержимое меню.
 */
export function withJsxContent<T extends WithItemsProps>(WrappedComponent: ComponentType<T>) {
    const WithJsxContent: FC<MenuJsxContentProps & Omit<T, 'items'>> = (props) => {
        const { items, children, ...otherProps } = props;
        const collection = useCollection<MixedItem>(props);

        return <WrappedComponent {...(otherProps as T)} items={items || collection} />;
    };

    WithJsxContent.displayName = `withJsxContent(${getDisplayName(WrappedComponent)})`;

    return WithJsxContent;
}

export * from './Group';
export * from './Item';
