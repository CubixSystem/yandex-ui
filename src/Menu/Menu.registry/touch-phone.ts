import { Registry } from '@bem-react/di';

import { cnMenu } from '../Menu';
import { MenuGroup } from '../Group/Menu-Group';
import { MenuItem } from '../Item/Menu-Item';
import { MenuText } from '../Text/Menu-Text';
import { MenuTitle } from '../Title/Menu-Title';
import { Icon } from '../../Icon/Icon';
import { withGlyphTypeCheck } from '../../Icon/_glyph/Icon_glyph_type-check';

export * from '.';

export const menuRegistry = new Registry({ id: cnMenu() });

menuRegistry
    .set('Group', MenuGroup)
    .set('Item', MenuItem)
    .set('Text', MenuText)
    .set('Title', MenuTitle)
    .set('Icon', withGlyphTypeCheck(Icon));
