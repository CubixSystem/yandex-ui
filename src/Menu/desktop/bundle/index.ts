import { compose, composeU, ExtractProps } from '@bem-react/core';

// base
import { Menu as MenuDesktop } from '../../Menu@desktop';
// _size
import { withSizeM } from '../../_size/Menu_size_m';
import { withSizeS } from '../../_size/Menu_size_s';
// _theme
import { withThemeNormal } from '../../_theme/Menu_theme_normal@desktop';
// _view
import { withViewDefault } from '../../_view/Menu_view_default@desktop';
// _width
import { withWidthAuto } from '../../_width/Menu_width_auto';
import { withWidthMax } from '../../_width/Menu_width_max';
// _jsxContent
import { withJsxContent } from '../../_jsxContent/Menu_jsxContent';

export * from '../../Menu@desktop';

export const Menu = compose(
    composeU(withSizeM, withSizeS),
    composeU(withWidthAuto, withWidthMax),
    withViewDefault,
    withThemeNormal,
)(withJsxContent(MenuDesktop));

export type IMenuProps = ExtractProps<typeof Menu>;
