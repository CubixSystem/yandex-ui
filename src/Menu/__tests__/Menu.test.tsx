import React, { createRef } from 'react';

import { createClientRender, screen, fireEvent } from '../../internal/testing';
import { delay } from '../../internal/utils';
import { Keys } from '../../lib/keyboard';
import { Menu as MenuCommon } from '../Menu';
import { Menu as MenuDesktop } from '../desktop';
import { Menu as MenuTouchPad } from '../touch-pad';
import { Menu as MenuTouchPhone } from '../touch-phone';
import { withJsxContent, Item, Group } from '../_jsxContent/Menu_jsxContent';

const platforms = [
    ['desktop', MenuDesktop],
    ['touch-pad', MenuTouchPad],
    ['touch-phone', MenuTouchPhone],
] as const;

describe.each(platforms)('Menu@%s', (platform, Menu) => {
    const render = createClientRender();

    test('должен вернуть полный шаблон компонента (snapshot)', () => {
        const items = [
            {
                title: 'test1',
                items: [
                    { id: 'item-id-1', value: '1', content: 'Тест1' },
                    { id: 'item-id-2', value: '2', content: 'Тест2' },
                ],
            },
            { id: 'item-id-3', value: '3', content: 'test2' },
        ];
        const { container } = render(<Menu items={items} className="mix" />);

        expect(container.firstChild).toMatchSnapshot();
    });

    test('должен устанавливать ссылку на корневой DOM элемент', async () => {
        const innerRef = createRef<HTMLElement>();
        render(<Menu items={[{ value: '1', content: 'Тест' }]} innerRef={innerRef} />);

        await delay(100);

        expect(innerRef.current).not.toBe(null);
    });

    test('должен быть установлен корректный displayName', () => {
        expect(MenuCommon.displayName).toBe('Menu');
    });

    test('должен проставить className Menu-Item_disabled и aria-selected=false при disabled всего меню', () => {
        const items = [
            { value: '1', content: 'Тест' },
            { value: '2', content: 'Тест2' },
            {
                title: 'test1',
                items: [
                    { value: '3', content: 'Тест3' },
                    { value: '4', content: 'Тест4' },
                ],
            },
        ];
        render(<Menu items={items} disabled />);

        const menuItems = screen.getAllByRole('menuitem');

        expect(menuItems).toHaveLength(4);

        menuItems.forEach((item) => {
            expect(item).toHaveClass('Menu-Item_disabled');
            expect(item).toHaveAttribute('aria-selected', 'false');
        });
    });

    test('должен проставляться className Menu-Item_disabled у item c disabled: true', () => {
        const items = [
            { value: '1', content: 'Тест' },
            { value: '2', content: 'Тест2' },
            {
                title: 'test1',
                items: [
                    { value: '3', content: 'Тест3', disabled: true },
                    { value: '4', content: 'Тест4' },
                ],
            },
        ];
        render(<Menu items={items} />);

        expect(screen.getByRole('menuitem', { name: 'Тест3' })).toHaveClass('Menu-Item_disabled');
    });

    test('должен вызывать onChange при нажатии на пункт меню', () => {
        const onChangeFn = jest.fn();
        const items = [
            { value: '1', content: 'Тест1' },
            { value: '2', content: 'Тест2' },
            {
                title: 'test1',
                items: [
                    { value: '3', content: 'Тест3' },
                    { value: '4', content: 'Тест4' },
                ],
            },
        ];
        render(<Menu items={items} onChange={onChangeFn} />);

        fireEvent.click(screen.getByText('Тест1'));

        expect(onChangeFn).toHaveBeenCalledTimes(1);
        expect(onChangeFn.mock.calls[0][0].target.value).toEqual('1');
    });

    test('должен при нажатии на пункт меню с multiselect вызывать onChange и в него передавать массив', () => {
        const onChangeFn = jest.fn();
        const items = [
            { value: '1', content: 'Тест1' },
            { value: '2', content: 'Тест2' },
            {
                title: 'test1',
                items: [
                    { value: '3', content: 'Тест3' },
                    { value: '4', content: 'Тест4' },
                ],
            },
        ];
        render(<Menu items={items} onChange={onChangeFn} value={[]} />);

        fireEvent.click(screen.getByText('Тест1'));

        expect(onChangeFn.mock.calls[0][0].target.value).toEqual(['1']);
    });

    test('должен корректно работать со строками, числами и объектами переданными по ссылке', () => {
        const onChangeFn = jest.fn();
        const refValue = { k: 1 };
        const items = [
            { value: 'str', content: 'Тест1' },
            { value: 2, content: 'Тест2' },
            { value: refValue, content: 'Тест3' },
        ];
        render(<Menu items={items} onChange={onChangeFn} />);

        fireEvent.click(screen.getByText('Тест1'));

        expect(onChangeFn.mock.calls[0][0].target.value).toBe('str');

        onChangeFn.mockReset();
        fireEvent.click(screen.getByText('Тест2'));

        expect(onChangeFn.mock.calls[0][0].target.value).toBe(2);

        onChangeFn.mockReset();
        fireEvent.click(screen.getByText('Тест3'));

        expect(onChangeFn.mock.calls[0][0].target.value).toBe(refValue);
    });

    test('должен переопределить элемент меню с помощью renderItem', () => {
        const items = [
            { value: 'testVal1', content: 'Тест1' },
            { value: 'testVal2', content: 'Тест2' },
        ];

        render(<Menu items={items} renderItem={(props) => <div data-testid="test-menuitem">{props.children}</div>} />);

        expect(screen.getAllByTestId('test-menuitem')).toHaveLength(2);
    });

    describe('_withJsxContent', () => {
        const MenuWithJsxContent = withJsxContent(Menu);

        test('должен проставляться дополнительный className у item', () => {
            render(
                <MenuWithJsxContent>
                    <Item className="TestMenuItem" value="1">
                        Тест1
                    </Item>
                </MenuWithJsxContent>,
            );

            expect(screen.getByRole('menuitem')).toHaveClass('TestMenuItem');
        });

        test('должен проставляться дополнительный className у group', () => {
            render(
                <MenuWithJsxContent>
                    <Group className="TestMenuGroup">
                        <Item value="testVal1">Тест1</Item>
                    </Group>
                </MenuWithJsxContent>,
            );

            expect(screen.getByRole('group')).toHaveClass('TestMenuGroup');
        });
    });

    test('должен прокидывать value в renderItem', () => {
        const renderItem = jest.fn(() => null);
        render(<Menu items={[{ value: 'test1', content: 'Тест1' }]} renderItem={renderItem} />);

        expect(renderItem).toHaveBeenCalledWith(expect.objectContaining({ value: 'test1' }), expect.anything());
    });

    if (platform === 'desktop') {
        test('должен перемещаться по активным пунктам меню при навигации клавиатурой', () => {
            const items = [
                { value: 'testVal1', content: 'Тест1', disabled: true },
                { value: 'testVal2', content: 'Тест2' },
                { value: 'testVal3', content: 'Тест3', disabled: true },
                { value: 'testVal4', content: 'Тест4' },
            ];
            render(<Menu items={items} onChange={() => null} value="" focused />);

            fireEvent.keyDown(document, { keyCode: Keys.UP });

            expect(screen.getByRole('option', { name: 'Тест4' })).toHaveClass('Menu-Item_hovered');

            fireEvent.keyDown(document, { keyCode: Keys.UP });

            expect(screen.getByRole('option', { name: 'Тест2' })).toHaveClass('Menu-Item_hovered');

            fireEvent.keyDown(document, { keyCode: Keys.DOWN });

            expect(screen.getByRole('option', { name: 'Тест4' })).toHaveClass('Menu-Item_hovered');

            fireEvent.keyDown(document, { keyCode: Keys.DOWN });

            expect(screen.getByRole('option', { name: 'Тест2' })).toHaveClass('Menu-Item_hovered');
        });
    }
});
