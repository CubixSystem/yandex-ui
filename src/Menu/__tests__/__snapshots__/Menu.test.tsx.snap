// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu@desktop должен вернуть полный шаблон компонента (snapshot) 1`] = `
<div
  aria-multiselectable="false"
  class="Menu mix"
>
  <div
    class="Menu-Group"
    role="group"
  >
    <div
      aria-hidden="true"
      class="Menu-Title"
    >
      test1
    </div>
    <div
      aria-selected="false"
      class="Menu-Item Menu-Item_type_menuitem"
      data-key="item-0"
      id="item-id-1"
      role="menuitem"
    >
      <span
        class="Menu-Text"
      >
        Тест1
      </span>
    </div>
    <div
      aria-selected="false"
      class="Menu-Item Menu-Item_type_menuitem"
      data-key="item-1"
      id="item-id-2"
      role="menuitem"
    >
      <span
        class="Menu-Text"
      >
        Тест2
      </span>
    </div>
  </div>
  <div
    aria-selected="false"
    class="Menu-Item Menu-Item_type_menuitem"
    data-key="item-2"
    id="item-id-3"
    role="menuitem"
  >
    <span
      class="Menu-Text"
    >
      test2
    </span>
  </div>
</div>
`;

exports[`Menu@touch-pad должен вернуть полный шаблон компонента (snapshot) 1`] = `
<div
  aria-multiselectable="false"
  class="Menu mix"
>
  <div
    class="Menu-Group"
    role="group"
  >
    <div
      aria-hidden="true"
      class="Menu-Title"
    >
      test1
    </div>
    <div
      aria-selected="false"
      class="Menu-Item Menu-Item_type_menuitem"
      data-key="item-0"
      id="item-id-1"
      role="menuitem"
    >
      <span
        class="Menu-Text"
      >
        Тест1
      </span>
    </div>
    <div
      aria-selected="false"
      class="Menu-Item Menu-Item_type_menuitem"
      data-key="item-1"
      id="item-id-2"
      role="menuitem"
    >
      <span
        class="Menu-Text"
      >
        Тест2
      </span>
    </div>
  </div>
  <div
    aria-selected="false"
    class="Menu-Item Menu-Item_type_menuitem"
    data-key="item-2"
    id="item-id-3"
    role="menuitem"
  >
    <span
      class="Menu-Text"
    >
      test2
    </span>
  </div>
</div>
`;

exports[`Menu@touch-phone должен вернуть полный шаблон компонента (snapshot) 1`] = `
<div
  aria-multiselectable="false"
  class="Menu mix"
>
  <div
    class="Menu-Group"
    role="group"
  >
    <div
      aria-hidden="true"
      class="Menu-Title"
    >
      test1
    </div>
    <div
      aria-selected="false"
      class="Menu-Item Menu-Item_type_menuitem"
      data-key="item-0"
      id="item-id-1"
      role="menuitem"
    >
      <span
        class="Menu-Text"
      >
        Тест1
      </span>
    </div>
    <div
      aria-selected="false"
      class="Menu-Item Menu-Item_type_menuitem"
      data-key="item-1"
      id="item-id-2"
      role="menuitem"
    >
      <span
        class="Menu-Text"
      >
        Тест2
      </span>
    </div>
  </div>
  <div
    aria-selected="false"
    class="Menu-Item Menu-Item_type_menuitem"
    data-key="item-2"
    id="item-id-3"
    role="menuitem"
  >
    <span
      class="Menu-Text"
    >
      test2
    </span>
  </div>
</div>
`;
